Remotion/Web/Core/ContentSecurityPolicy/CspEnabledHtmlTextWriter.cs

```diff
@@ -0,0 +1,113 @@
+﻿// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+// 
+// The re-motion Core Framework is free software; you can redistribute it
+// and/or modify it under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation; either version 2.1 of the
+// License, or (at your option) any later version.
+// 
+// re-motion is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+#if !NETFRAMEWORK
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Web.UI;
+using Remotion.Utilities;
+using Remotion.Web.UI;
+
+namespace Remotion.Web.ContentSecurityPolicy
+{
+  /// <summary>
+  /// Represents <see cref="HtmlTextWriter"/> which implements Content-Security-Policy
+  /// </summary>
+  public class CspEnabledHtmlTextWriter : HtmlTextWriter
+  {
+    private static readonly IReadOnlyDictionary<string, string> s_supportedEvents =
+        new Dictionary<string, string>
+        {
+          { "onclick", "click" },
+          { "onchange", "change" },
+          { "onmouseover", "mouseover" },
+          { "onmouseout", "mouseout" },
+          { "onkeyup", "keyup" },
+          { "onkeydown", "keydown" },
+          { "onkeypress", "keypress" }
+        };
+
+    private readonly List<(string type, string value)> _registeredEvents = new();
+    private readonly INonceGenerator _nonceGenerator;
+    private readonly ISmartPage _page;
+    private readonly string _requestNonce;
+
+    public CspEnabledHtmlTextWriter (ISmartPage page, TextWriter writer, INonceGenerator nonceGenerator, string requestNonce)
+        : base(writer)
+    {
+      ArgumentUtility.CheckNotNull("page", page);
+      ArgumentUtility.CheckNotNull("writer", writer);
+      ArgumentUtility.CheckNotNull("nonceGenerator", nonceGenerator);
+      ArgumentUtility.CheckNotEmpty("requestNonce", requestNonce);
+
+      _page = page;
+      _nonceGenerator = nonceGenerator;
+      _requestNonce = requestNonce;
+    }
+
+    public override void RenderBeginTag (HtmlTextWriterTag tagKey)
+    {
+      if (tagKey == HtmlTextWriterTag.Script)
+      {
+        AddAttribute("nonce", _requestNonce);
+      }
+
+      if (_registeredEvents.Count > 0)
+      {
+        var eventTargetID = _nonceGenerator.GenerateAlphaNumericNonce();
+        AddAttribute("data-inline-event-target", eventTargetID);
+
+        foreach (var registeredEvent in _registeredEvents)
+        {
+
+          var script = $"SmartPageContext.Instance.RegisterInlineEvent(\"{eventTargetID}\", \"{registeredEvent.type}\", function (event){{{registeredEvent.value}}});";
```

Comment:
I always use single quotes for rendered Javascript. that way, I don't need to escape the quotes in the script string.